
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.32">
    
    
      
        <title>Documentation for oarfish</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#oarfish-transcript-quantification-from-long-read-rna-seq-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="Documentation for oarfish" class="md-header__button md-logo" aria-label="Documentation for oarfish" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Documentation for oarfish
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              oarfish: transcript quantification from long-read RNA-seq data
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="Documentation for oarfish" class="md-nav__button md-logo" aria-label="Documentation for oarfish" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Documentation for oarfish
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    oarfish: transcript quantification from long-read RNA-seq data
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    oarfish: transcript quantification from long-read RNA-seq data
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#precompiled-binaries" class="md-nav__link">
    <span class="md-ellipsis">
      Precompiled binaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-cargo" class="md-nav__link">
    <span class="md-ellipsis">
      Using cargo
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bioconda" class="md-nav__link">
    <span class="md-ellipsis">
      Bioconda
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-usage" class="md-nav__link">
    <span class="md-ellipsis">
      Basic usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Usage examples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usage examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#aligment-mode-example" class="md-nav__link">
    <span class="md-ellipsis">
      aligment-mode example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mode-example" class="md-nav__link">
    <span class="md-ellipsis">
      read-mode example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-to-oarfish" class="md-nav__link">
    <span class="md-ellipsis">
      Input to oarfish
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input to oarfish">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-based-input" class="md-nav__link">
    <span class="md-ellipsis">
      Read-based input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alignmment-based-input" class="md-nav__link">
    <span class="md-ellipsis">
      Alignmment-based input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-notes-on-oarfish-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Other notes on oarfish parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#notes-about-single-cell-mode" class="md-nav__link">
    <span class="md-ellipsis">
      Notes about single-cell mode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inferential-replicates" class="md-nav__link">
    <span class="md-ellipsis">
      Inferential Replicates
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#precompiled-binaries" class="md-nav__link">
    <span class="md-ellipsis">
      Precompiled binaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-cargo" class="md-nav__link">
    <span class="md-ellipsis">
      Using cargo
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bioconda" class="md-nav__link">
    <span class="md-ellipsis">
      Bioconda
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-usage" class="md-nav__link">
    <span class="md-ellipsis">
      Basic usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Usage examples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usage examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#aligment-mode-example" class="md-nav__link">
    <span class="md-ellipsis">
      aligment-mode example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mode-example" class="md-nav__link">
    <span class="md-ellipsis">
      read-mode example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-to-oarfish" class="md-nav__link">
    <span class="md-ellipsis">
      Input to oarfish
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input to oarfish">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-based-input" class="md-nav__link">
    <span class="md-ellipsis">
      Read-based input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alignmment-based-input" class="md-nav__link">
    <span class="md-ellipsis">
      Alignmment-based input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-notes-on-oarfish-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Other notes on oarfish parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#notes-about-single-cell-mode" class="md-nav__link">
    <span class="md-ellipsis">
      Notes about single-cell mode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inferential-replicates" class="md-nav__link">
    <span class="md-ellipsis">
      Inferential Replicates
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="oarfish-transcript-quantification-from-long-read-rna-seq-data">oarfish: transcript quantification from long-read RNA-seq data</h1>
<p><code>oarfish</code> is a program, written in Rust (https://www.rust-lang.org/), for quantifying transcript-level expression from long-read (i.e. Oxford nanopore cDNA and direct RNA and PacBio) sequencing technologies. <code>oarfish</code> requires a sample of sequencing reads aligned to the <em>transcriptome</em> (currntly not to the genome). It handles multi-mapping reads through the use of probabilistic allocation via an expectation-maximization (EM) algorithm.</p>
<p>It optionally employs many filters to help discard alignments that may reduce quantification accuracy. Currently, the set of filters applied in <code>oarfish</code> are directly derived from the <a href="https://github.com/a-slide/NanoCount"><code>NanoCount</code></a><sup id="fnref:Gleeson"><a class="footnote-ref" href="#fn:Gleeson">1</a></sup> tool; both the filters that exist, and the way their values are set (with the exception of the <code>--three-prime-clip</code> filter, which is not set by default in <code>oarfish</code> but is in <code>NanoCount</code>).</p>
<p>Additionally, <code>oarfish</code> provides options to make use of coverage profiles derived from the aligned reads to improve quantification accuracy. The use of this coverage model is enabled with the <code>--model-coverage</code> flag. You can read more about <code>oarfish</code><sup id="fnref:preprint"><a class="footnote-ref" href="#fn:preprint">2</a></sup> in the <a href="https://www.biorxiv.org/content/10.1101/2024.02.28.582591v1">preprint</a>. Please cite the preprint if you use <code>oarfish</code> in your work or analysis.</p>
<p>Also, please note that <code>oarfish</code> is scientific software in active development. Therefore, please check the <a href="https://github.com/COMBINE-lab/oarfish/releases">GitHub Release</a> page to make sure that you are using the latest version</p>
<h2 id="installation">Installation</h2>
<p><code>oarfish</code> can be installed in a variety of ways.</p>
<h3 id="precompiled-binaries">Precompiled binaries</h3>
<p>Binaries are available via <a href="https://github.com/COMBINE-lab/oarfish/releases">GitHub Releases</a>.</p>
<p>You can quickly install the latest release using the following helper script:</p>
<pre><code class="language-sh">curl --proto '=https' --tlsv1.2 -LsSf https://github.com/COMBINE-lab/oarfish/releases/latest/download/oarfish-installer.sh | sh
</code></pre>
<h3 id="using-cargo">Using <code>cargo</code></h3>
<p>If you have <code>cargo</code> installed, you can install <code>oarfish</code> directly from the source code:</p>
<pre><code class="language-sh">cargo install oarfish
</code></pre>
<p>You can find the crate on <a href="https://crates.io/crates/oarfish">crates.io</a>.</p>
<h3 id="bioconda">Bioconda</h3>
<p><code>oarfish</code> is available via <a href="https://anaconda.org/bioconda/oarfish">Bioconda</a>:</p>
<pre><code class="language-sh">conda install -c bioconda oarfish
</code></pre>
<h2 id="basic-usage">Basic usage</h2>
<p>The usage can be provided by passing <code>-h</code> at the command line.</p>
<pre><code>A fast, accurate and versatile tool for long-read transcript quantification.

Usage: oarfish [OPTIONS] --output &lt;OUTPUT&gt; &lt;--alignments &lt;ALIGNMENTS&gt;|--reads &lt;READS&gt;&gt;

Options:
      --quiet
          be quiet (i.e. don't output log messages that aren't at least warnings)
      --verbose
          be verbose (i.e. output all non-developer logging messages)
  -o, --output &lt;OUTPUT&gt;
          location where output quantification file should be written
      --single-cell
          input is assumed to be a single-cell BAM and to have the `CB:z` tag for all read records
  -j, --threads &lt;THREADS&gt;
          maximum number of cores that the oarfish can use to obtain binomial probability [default: 1]
      --num-bootstraps &lt;NUM_BOOTSTRAPS&gt;
          number of bootstrap replicates to produce to assess quantification uncertainty [default: 0]
  -h, --help
          Print help
  -V, --version
          Print version

alignment mode:
  -a, --alignments &lt;ALIGNMENTS&gt;  path to the file containing the input alignments

raw read mode:
      --reads &lt;READS&gt;          path to the file containing the input reads
      --reference &lt;REFERENCE&gt;  path to the file containing the reference transcriptome (or existing index) against which to map
      --index-out &lt;INDEX_OUT&gt;  path where minimap2 index will be written (if provided)
      --seq-tech &lt;SEQ_TECH&gt;    sequencing technology in which to expect reads if using mapping based mode [possible values: ont-cdna, ont-drna, pac-bio, pac-bio-hifi]
      --best-n &lt;BEST_N&gt;        maximum number of secondary mappings to consider when mapping reads to the transcriptome [default: 100]

filters:
      --filter-group &lt;FILTER_GROUP&gt;
          [possible values: no-filters, nanocount-filters]
  -t, --three-prime-clip &lt;THREE_PRIME_CLIP&gt;
          maximum allowable distance of the right-most end of an alignment from the 3' transcript end [default: *4294967295]
  -f, --five-prime-clip &lt;FIVE_PRIME_CLIP&gt;
          maximum allowable distance of the left-most end of an alignment from the 5' transcript end [default: *4294967295]
  -s, --score-threshold &lt;SCORE_THRESHOLD&gt;
          fraction of the best possible alignment score that a secondary alignment must have for consideration [default: *0.95]
  -m, --min-aligned-fraction &lt;MIN_ALIGNED_FRACTION&gt;
          fraction of a query that must be mapped within an alignemnt to consider the alignemnt valid [default: *0.5]
  -l, --min-aligned-len &lt;MIN_ALIGNED_LEN&gt;
          minimum number of nucleotides in the aligned portion of a read [default: *50]
  -d, --strand-filter &lt;STRAND_FILTER&gt;
          only alignments to this strand will be allowed; options are (fw /+, rc/-, or both/.) [default: .]

coverage model:
      --model-coverage         apply the coverage model
  -b, --bin-width &lt;BIN_WIDTH&gt;  width of the bins used in the coverage model [default: 100]

EM:
      --max-em-iter &lt;MAX_EM_ITER&gt;
          maximum number of iterations for which to run the EM algorithm [default: 1000]
      --convergence-thresh &lt;CONVERGENCE_THRESH&gt;
          maximum number of iterations for which to run the EM algorithm [default: 0.001]
  -q, --short-quant &lt;SHORT_QUANT&gt;
          location of short read quantification (if provided)
</code></pre>
<h2 id="usage-examples">Usage examples</h2>
<p>Assume that you have ONT cDNA sequencing reads in a file named <code>sample1_reads.tar.gz</code>, and you'd like to quantify the transcripts in a <em>transcriptome</em> reference in the file <code>transcripts.fa</code>.
To accomplish this with oarfish, you can use either <a href="./#alignment-based-input">alignment-based</a> or <a href="./#read-based-input">read-based</a> mode.  Here we give a brief example
of each.  To use alignment-based mode, we assume you have <a href="https://github.com/lh3/minimap2">minimap2</a> and <a href="http://www.htslib.org/">samtools</a> installed.  </p>
<h3 id="aligment-mode-example">aligment-mode example</h3>
<p>You can quantify the transcript abundances in this sample using the following commands:</p>
<pre><code>$ minimap2 -t 16 -ax map-ont transcripts.fa sample1_reads.tar.gz | samtools view -@4 -b -o alignments.bam
$ oarfish -j 16 -a alignments.bam -o sample1 --filter-group no-filters --model-coverage
</code></pre>
<p>This will produce several output files, as described <a href="./#output">below</a>.</p>
<h3 id="read-mode-example">read-mode example</h3>
<p>In read-based mode, you can quantify the transcript abundances in this sample using the following commands:</p>
<pre><code>$ oarfish -j 16 --reads sample1_reads.tar.gz --reference transcripts.fa --seq-tech ont-cdna -o sample1 --filter-group no-filters --model-coverage
</code></pre>
<p>If you are going to quantify more than one sample against these reference transcripts, it makes sense to save the minimap2 index that the above
command creates.  This can be done using the following command:</p>
<pre><code>$ oarfish -j 16 --reads sample1_reads.tar.gz --reference transcripts.fa --index-out transcripts.mmi --seq-tech ont-cdna -o sample1 --filter-group no-filters --model-coverage
</code></pre>
<p>Then, in subsequent runs (say when quantifying <code>sample2_reads.tar.gz</code>), you can directly provide the <code>minimap2</code> index in place of the reference to
speed up quantification.  That command would look like the following:</p>
<pre><code>$ oarfish -j 16 --reads sample2_reads.tar.gz --reference transcripts.mmi --seq-tech ont-cdna -o sample2 --filter-group no-filters --model-coverage
</code></pre>
<p>As with alignment-based mode, these commands will produce several output files, as described <a href="./#output">below</a>.</p>
<h2 id="input-to-oarfish">Input to <code>oarfish</code></h2>
<p><code>Oarfish</code> can accept as input either a <code>bam</code> file containing reads aligned to the transcriptome as specified <a href="./#alignment-based-input">below</a>, or
raw sequencing reads themselves (along with a reference transcriptome), which are then mapped to the reference using <a href="https://github.com/jguhlin/minimap2-rs">minimap2-rs</a>
and subsequently processed with <code>oarfish</code>.  With equivalent alignment options, the results of these input modes should be equivalent, so which to use is therefore
based on the preference of the user.</p>
<h3 id="read-based-input">Read-based input</h3>
<p>The read-based input mode takes as input a reference (specified with the <code>--reference</code> argument), which can be either a <code>FASTA</code> file containing a transcriptome reference
or an pre-build <code>minimap2</code> index, as well as a set of reads (specified with the <code>--reads</code> argument), and a <code>--seq-tech</code> argument specifying the sequencing technology 
type of the reads to be mapped.</p>
<p>The mapping between the potential values that can be passed to <code>oarfish</code>'s <code>--seq-tech</code> argument and the <code>minimap2</code> presets is as follows:</p>
<ul>
<li><code>oarfish</code> seq-tech <code>ont-cdna</code> corresponds to <code>minimap2</code> preset <code>map-ont</code></li>
<li><code>oarfish</code> seq-tech <code>ont-drna</code> corresponds to <code>minimap2</code> preset <code>map-ont</code></li>
<li><code>oarfish</code> seq-tech <code>pac-bio</code> corresponds to <code>minimap2</code> preset <code>map-pb</code></li>
<li><code>oarfish</code> seq-tech <code>pac-bio-hifi</code> corresponds to <code>minimap2</code> preset <code>map-hifi</code></li>
</ul>
<p>Given these inputs, <code>oarfish</code> will either load the pre-built <code>minimap2</code> index, or build one according to the parameter specified by <code>--seq-tech</code>, and will then align
the reads to this index using <a href="https://github.com/jguhlin/minimap2-rs"><code>minimap2-rs</code></a>.  Optionally, the maximum multimapping rate (i.e. the number of secondary alignments 
corresponding to the <code>minimap2</code> parameter <code>-N</code>) can be specified with the command line parameter <code>--best-n</code>. The default value of this parameter is 100.</p>
<h3 id="alignmment-based-input">Alignmment-based input</h3>
<p>In alignment-based mode, <code>oarfish</code> processes pre-computed alignments of hte read to the transcriptome. The input should be a <code>bam</code> format file, with reads aligned using <a href="https://github.com/lh3/minimap2"><code>minimap2</code></a> against the <em>transcriptome</em>. That is, <code>oarfish</code> does not currently handle spliced alignment to the genome. Further, the output alignments should be name sorted (the default order produced by <code>minimap2</code> should be fine). <em>Specifically</em>, <code>oarfish</code> relies on the existence of the <code>AS</code> tag in the <code>bam</code> records that encodes the alignment score in order to obtain the score for each alignment (which is used in probabilistic read assignment), and the score of the best alignment, overall, for each read. ### Choosing <code>minimap2</code> alignment options Since the purpose of <code>oarfish</code> is to estimate transcript abundance from a collection of alignments to the target transcriptome, it is important that the alignments are generated in a fashion that is compatible with this goal.  Primarily, this means that the aligner should be configured to report as many optimal (and near-optimal) alignments as exist, so that <code>oarfish</code> can observe all of this information and determine how to allocate reads to transcripts.  We recommend using the following options with <code>minimap2</code> when aligning data for later processing by <code>oarfish</code> * For ONT data (either dRNA or cDNA): please use the flags <code>--eqx -N 100 -ax map-ont</code> For PacBio data: please use the flags <code>--eqx -N 100 -ax pacbio</code> <strong>Note (1)</strong>: It may be worthwile using an even larger <code>N</code> value (e.g. the <a href="https://www.biorxiv.org/content/10.1101/2024.04.13.589356v1.full">TranSigner manuscript</a> recommends <code>-N 181</code>). A larger value should not diminish the accuracy of <code>oarfish</code>, but it may make alignment take longer and produce a larger <code>bam</code> file.</p>
<p><strong>Note (2)</strong>: For very high quality PacBio data, it may be most appropriate to use the <code>-ax map-hifi</code> flag in place of <code>-ax pacbio</code>.  We are currently evaluating the effect of this option, and also welcome feedback if you have experiences to share on the use of data aligned with these different flags with <code>oarfish</code>.</p>
<h3 id="other-notes-on-oarfish-parameters">Other notes on <code>oarfish</code> parameters</h3>
<p>The parameters above should be explained by their relevant help option, but the <code>-d</code>/<code>--strand-filter</code> is worth noting explicitly. By default, alignments to both strands of a transcript will be considered valid.  You can use this option to allow only alignments in the specified orientation; for example <code>-d fw</code> will allow only alignments in the forward orientation and <code>-d rc</code> will allow only alignments in the reverse-complement orientation and <code>-d both</code> (the default) will allow both.  The <code>-d</code> filter, if explicitly provided, overrides the orientation filter in any provided "filter group" so e.g. passing <code>--filter-group no-filters -d fw</code> will disable other filters, but will still only admit alignments in the forward orientation.</p>
<p><strong>In general</strong>, if you apply a <code>filter-group</code>, the group options will be applied first and then any explicitly provided options given will override the corresponding option in the <code>filter-group</code>.</p>
<h2 id="notes-about-single-cell-mode">Notes about single-cell mode</h2>
<p>Starting with version 0.6.1 <code>oarfish</code> incorporates the first single-cell quantification capabilities. Given a <code>bam</code> file, <strong>collated by cell barcode and with already (UMI) deduplicated reads</strong>, this mode, enabled with the <code>--single-cell</code> flag, will allow <code>oarfish</code> to produce a single-cell quantification matrix. Currently, this mode can not be used with read-based mode, and the input <code>bam</code> file should be properly formatted for this purpose. </p>
<p><strong>Formatting requirements of BAM input in single-cell mode</strong>: All alignment records for the same cell barcode should be adjacent in the <code>bam</code> file, and a count will be obtained for each read record, so UMI de-duplication should have been performed if those are the counts you want. In the future, counting UMIs directly may be supported, and some of these other restrictions may be lifted.</p>
<h2 id="inferential-replicates">Inferential Replicates</h2>
<p><code>oarfish</code> has the ability to compute <a href="https://academic.oup.com/nar/article/47/18/e105/5542870"><em>inferential replicates</em></a> of its quantification estimates. This is performed by bootstrap sampling of the original read mappings, and subsequently performing inference under each resampling.  These inferential replicates allow assessing the variance of the point estimate of transcript abundance, and can lead to improved differential analysis at the transcript level, if using a differential testing tool that takes advantage of this information. The generation of inferential replicates is controlled by the <code>--num-bootstraps</code> argument to <code>oarfish</code>.  The default value is <code>0</code>, meaning that no inferential replicates are generated.  If you set this to some value greater than <code>0</code>, the the requested number of inferential replicates will be generated. It is recommended, if generating inferential replicates, to run <code>oarfish</code> with multiple threads, since replicate generation is highly-parallelized. Finally, if replicates are generated, they are written to a <a href="https://parquet.apache.org/"><code>Parquet</code></a>, starting with the specified output stem and ending with <code>infreps.pq</code>.</p>
<h2 id="output">Output</h2>
<p>The <code>--output</code> option passed to <code>oarfish</code> corresponds to a path prefix (this prefix can contain the path separator character and if it refers to a directory that does not yeat exist, that directory will be created). Based on this path prefix, say <code>P</code>, <code>oarfish</code> will create 2 files:</p>
<ul>
<li><code>P.meta_info.json</code> - a JSON format file containing information about relevant parameters with which <code>oarfish</code> was run, and other relevant inforamtion from the processed sample apart from the actual transcript quantifications.</li>
<li><code>P.quant</code> - a tab separated file listing the quantified targets, as well as information about their length and other metadata. The <code>num_reads</code> column provides the estimate of the number of reads originating from each target.</li>
<li><code>P.infreps.pq</code> - a <a href="https://parquet.apache.org/"><code>Parquet</code></a> table where each row is a transcript and each column is an inferential replicate, containing the estimated counts for each transcript under each computed inferential replicate.</li>
<li><code>P.ambig_info.tsv</code> - a tab separated file listing, for each transcript (in the same order in which they appear in <code>P.quant</code>) the number of uniquely mapped, ambiguously mapped, and total reads.  The quantification estimate for each transcript, in general, should reside between the number of uniquely aligned reads and the total number of reads (i.e. these provide, respectively lower and upper bounds for the number of reads assigned to each transcript).  Note that the total in this file is the total number of reads that align to this transcript with a sufficiently high alignment score --- it is <em>not</em>, in general, an estimate of the number of reads originating from this transcript as many of those reads can be multimapping and, in fact, potentially better described by other transcripts.</li>
</ul>
<h2 id="references">References</h2>
<div class="footnote">
<hr />
<ol>
<li id="fn:Gleeson">
<p>Josie Gleeson, Adrien Leger, Yair D J Prawer, Tracy A Lane, Paul J Harrison, Wilfried Haerty, Michael B Clark, Accurate expression quantification from nanopore direct RNA sequencing with NanoCount, Nucleic Acids Research, Volume 50, Issue 4, 28 February 2022, Page e19, <a href="https://doi.org/10.1093/nar/gkab1129">https://doi.org/10.1093/nar/gkab1129</a>&#160;<a class="footnote-backref" href="#fnref:Gleeson" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
<li id="fn:preprint">
<p>Zahra Zare Jousheghani, Rob Patro. Oarfish: Enhanced probabilistic modeling leads to improved accuracy in long read transcriptome quantification, bioRxiv 2024.02.28.582591; doi: <a href="https://doi.org/10.1101/2024.02.28.582591">https://doi.org/10.1101/2024.02.28.582591</a>&#160;<a class="footnote-backref" href="#fnref:preprint" title="Jump back to footnote 2 in the text">&#8617;</a></p>
</li>
</ol>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.471ce7a9.min.js"></script>
      
        <script src="javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>